### PR TITLE
if function is added

### PR DESCRIPTION
**Pull request by rpichaimani**

### CODE REVIEW

File: src/main/java/org/valdroz/vscript/BaseNode.java [hunk 1]
```diff
@@ -341,17 +341,32 @@ public Variant execute(VariantContainer variantContainer) {
...
+                    if ((result.asBoolean() && result.isBoolean() == true)) {
+                        result = Variant.fromString(params.get(1).name);
+                    } else {
+                        result = Variant.fromString(params.get(2).name);
                     }
                 }
-                result = function.execute(lvc);
-                if ((result == null || result.isNull()) && (valueSubstitution != null)) {
-                    result = valueSubstitution.execute(variantContainer);
+                else{
```
Comment (valdroz) [inline src/main/java/org/valdroz/vscript/BaseNode.java]: Code formatting issue. Missing space
```suggestion
                else {
```

File: src/main/java/org/valdroz/vscript/BaseNode.java [hunk 2]
```diff
@@ -341,17 +341,32 @@ public Variant execute(VariantContainer variantContainer) {
                 }
                 LocalVariantContainer lvc = new LocalVariantContainer(variantContainer);
                 List<String> parameterNames = function.getParameterNames();
-                for (int pidx = 0;
-                     pidx < Math.min(params.size(), parameterNames.size());
-                     ++pidx) {
-                    String fpn = parameterNames.get(pidx);
-                    if (fpn.length() > 0) {
-                        lvc.setVariant(fpn, params.get(pidx).execute(variantContainer));
+                if(getName().equalsIgnoreCase(NT_FUNCTION_IF)) {
```
Comment (valdroz) [inline src/main/java/org/valdroz/vscript/BaseNode.java]: 1) Code formatting issueL Missing space after `if`:
2) Can a handling for function `if` be implemented as separate `case NT_MF_IF`?

File: src/main/java/org/valdroz/vscript/BaseNode.java [hunk 3]
```diff
@@ -341,17 +341,32 @@ public Variant execute(VariantContainer variantContainer) {
...
+                    if (parameterNames.size() != 3) {
+                        throw new EvaluationException("Illegal Number of Paramters.");
+                    }
+                    lvc.setVariant(parameterNames.get(0), params.get(0).execute(variantContainer));
+
+                    result = function.execute(lvc);
+                    if ((result.asBoolean() && result.isBoolean() == true)) {
+                        result = Variant.fromString(params.get(1).name);
```
Comment (valdroz) [inline src/main/java/org/valdroz/vscript/BaseNode.java]: how will this work when expression is
```
a=10;b=a*2; if(true, a, b)
```
? Will we get `10` as a result?

File: src/main/java/org/valdroz/vscript/BaseNode.java (hunk 1 above)
Comment (rpichaimani) [reply src/main/java/org/valdroz/vscript/BaseNode.java]: fixed formatting issue.

File: src/main/java/org/valdroz/vscript/BaseNode.java (hunk 2 above)
Comment (rpichaimani) [reply src/main/java/org/valdroz/vscript/BaseNode.java]: introduced `NT_MF_IF` implementation

File: src/main/java/org/valdroz/vscript/BaseNode.java (hunk 3 above)
Comment (rpichaimani) [reply src/main/java/org/valdroz/vscript/BaseNode.java]: added test case on this scenario to cover.

File: src/main/java/org/valdroz/vscript/BaseNode.java [hunk 4]
```diff
@@ -353,9 +353,40 @@ public Variant execute(VariantContainer variantContainer) {
...
                 }
+
                 result = Variant.sanitize(result);
             }
             break;
+            case NT_MF_IF: {
+                if (parentRunBlock == null) {
+                    throw new UndefinedFunction(getName());
+                }
+                AbstractFunction function = parentRunBlock.resolveFunction(getName());
```
Comment (valdroz) [inline src/main/java/org/valdroz/vscript/BaseNode.java]: This is not necessary as the node it self is a `if` function

File: src/main/java/org/valdroz/vscript/BaseNode.java [hunk 5]
```diff
@@ -353,9 +353,40 @@ public Variant execute(VariantContainer variantContainer) {
...
+                if (function == null) {
+                    throw new UndefinedFunction(getName());
+                }
+                LocalVariantContainer lvc = new LocalVariantContainer(variantContainer);
+                List<String> parameterNames = function.getParameterNames();
+                if (parameterNames.size() != 3) {
+                    throw new EvaluationException("Illegal Number of Paramters.");
```
Comment (valdroz) [inline src/main/java/org/valdroz/vscript/BaseNode.java]: Error message does not provide sufficient detail to consumer to understand what is wrong`Illegal Number of Paramters.`

File: src/main/java/org/valdroz/vscript/BaseNode.java (hunk 4 above)
Comment (valdroz) [reply src/main/java/org/valdroz/vscript/BaseNode.java]: purpose of `parentRunBlock` is to resolve function that are not native to vscript engine

File: src/main/java/org/valdroz/vscript/BaseNode.java [hunk 6]
```diff
@@ -353,9 +353,40 @@ public Variant execute(VariantContainer variantContainer) {
                 if ((result == null || result.isNull()) && (valueSubstitution != null)) {
                     result = valueSubstitution.execute(variantContainer);
                 }
+
                 result = Variant.sanitize(result);
             }
             break;
+            case NT_MF_IF: {
```
Comment (valdroz) [inline src/main/java/org/valdroz/vscript/BaseNode.java]: `case NT_MF_IF:` could be modeled after `case NT_MF_TO_ARRAY:`

File: src/main/java/org/valdroz/vscript/EquationParser.java [hunk 7]
```diff
@@ -454,13 +455,18 @@ private BaseNode parseAsIdentifiable() {
...
             if (isFunction) {
-                if (mathFuncOpCode > 0) {
+                if (mathFuncOpCode > 0 &&  mathFuncOpCode != 170){
                     node = newNode()
                             .withNodeOperation(mathFuncOpCode);
                     if (parseParamsFor(node, '(', ')', ',') == 0) {
                         node.addParameterNode(Constants.C_NULL);
                     }
-                } else {
+                } else if (mathFuncOpCode == 170) {
```
Comment (valdroz) [inline src/main/java/org/valdroz/vscript/EquationParser.java]: `else if (mathFuncOpCode == 170)`  block is not reuquired

File: src/main/java/org/valdroz/vscript/EquationParser.java [hunk 8]
```diff
@@ -454,13 +455,18 @@ private BaseNode parseAsIdentifiable() {
             int mathFuncOpCode = determineBuildinFuncCode(word);
 
             if (isFunction) {
-                if (mathFuncOpCode > 0) {
+                if (mathFuncOpCode > 0 &&  mathFuncOpCode != 170){
```
Comment (valdroz) [inline src/main/java/org/valdroz/vscript/EquationParser.java]: `&& mathFuncOpCode != 170` is not required

File: src/test/java/org/valdroz/vscript/examples/ExampleCustomFunction.java [hunk 9]
```diff
@@ -17,5 +16,27 @@ public static void main(String[] args) {
                 .eval(variantContainer);
 
         assertThat(result.asNumeric().doubleValue(), is(14.0));
+
+        masterRunBlock.registerFunction("if(condition, first, second)",
```
Comment (valdroz) [inline src/test/java/org/valdroz/vscript/examples/ExampleCustomFunction.java]: `masterRunBlock.registerFunction("if(condition, first, second)"` should not be used as the function `if` is supposed to be handled natively by `vscript` just like  other inbuilt functions, e.g. `to_array`

**Review by valdroz — CHANGES_REQUESTED** [review]

File: src/main/java/org/valdroz/vscript/BaseNode.java (hunk 4 above)
Comment (rpichaimani) [reply src/main/java/org/valdroz/vscript/BaseNode.java]: updated the `NT_MF_IF` block.

File: src/main/java/org/valdroz/vscript/BaseNode.java (hunk 6 above)
Comment (rpichaimani) [reply src/main/java/org/valdroz/vscript/BaseNode.java]: `NT_MF_IF` modeled after `NT_MF_TO_ARRAY`

File: src/main/java/org/valdroz/vscript/EquationParser.java (hunk 8 above)
Comment (rpichaimani) [reply src/main/java/org/valdroz/vscript/EquationParser.java]: removed condition `&& mathFuncOpCode != 170`

File: src/test/java/org/valdroz/vscript/examples/ExampleCustomFunction.java (hunk 9 above)
Comment (rpichaimani) [reply src/test/java/org/valdroz/vscript/examples/ExampleCustomFunction.java]: updated. We can discuss tomorrow if the changes are still incorrect. Please let me know. Thanks

File: src/main/java/org/valdroz/vscript/BaseNode.java (hunk 5 above)
Comment (rpichaimani) [reply src/main/java/org/valdroz/vscript/BaseNode.java]: updated Error message.

File: src/main/java/org/valdroz/vscript/EquationParser.java (hunk 7 above)
Comment (rpichaimani) [reply src/main/java/org/valdroz/vscript/EquationParser.java]: removed `else if (mathFuncOpCode == 170)` block

File: src/main/java/org/valdroz/vscript/BaseNode.java [hunk 10]
```diff
@@ -330,7 +330,17 @@ public Variant execute(VariantContainer variantContainer) {
                 result = Variant.fromArray(arrItems);
             }
             break;
-
+            case NT_MF_IF: {
+                List<Variant> parameterItems = Lists.newArrayList();
+                if (params != null) {
```
Comment (valdroz) [inline src/main/java/org/valdroz/vscript/BaseNode.java]: Implemention of `if` function should evaluate first parameter, and then based on the result `true`/`false`, evaluate and respond with result from either second or third parameter. All parameters should never be evaluated.

File: src/main/java/org/valdroz/vscript/BaseNode.java [hunk 11]
```diff
@@ -330,7 +330,17 @@ public Variant execute(VariantContainer variantContainer) {
...
+                    params.forEach(itemNode -> parameterItems.add(itemNode.execute(variantContainer)));
+                }
+                if (parameterItems.size() != 3) {
+                    throw new EvaluationException("Number of Parameters must be 3. Sample if(true, a, b)");
+                }
+                result = Variant.fromIf(parameterItems);
```
Comment (valdroz) [inline src/main/java/org/valdroz/vscript/BaseNode.java]: Variant shall not implement `if` function

File: src/main/java/org/valdroz/vscript/Variant.java [hunk 12]
```diff
@@ -120,6 +120,10 @@ public static Variant fromArray(List<Variant> value) {
         return new ArrayVariant(value);
     }
 
+    public static Variant fromIf(List<Variant> value) {
```
Comment (valdroz) [inline src/main/java/org/valdroz/vscript/Variant.java]: Variant shall not implement functions

File: src/main/java/org/valdroz/vscript/BaseNode.java [hunk 13]
```diff
@@ -330,7 +330,17 @@ public Variant execute(VariantContainer variantContainer) {
                 result = Variant.fromArray(arrItems);
             }
             break;
-
+            case NT_MF_IF: {
+                List<Variant> parameterItems = Lists.newArrayList();
+                if (params != null) {
+                    params.forEach(itemNode -> parameterItems.add(itemNode.execute(variantContainer)));
+                }
+                if (parameterItems.size() != 3) {
```
Comment (valdroz) [inline src/main/java/org/valdroz/vscript/BaseNode.java]: Validation should be done before any parameter nodes are evaluated

**Review by valdroz — CHANGES_REQUESTED** [review]

File: src/main/java/org/valdroz/vscript/BaseNode.java [hunk 14]
```diff
@@ -330,7 +330,17 @@ public Variant execute(VariantContainer variantContainer) {
                 result = Variant.fromArray(arrItems);
             }
             break;
-
+            case NT_MF_IF: {
```
Comment (valdroz) [inline src/main/java/org/valdroz/vscript/BaseNode.java:333]:
```suggestion
                case NT_MF_IF:
                if (params == null || params.size() < 3) {
                    throw new EvaluationException("Function `if` takes 3 parameters. E.g. if(true, \"it is true\", \"it is false\")");
                }
                
                result = (params.get(0).execute(variantContainer).asBoolean()) ?
                        params.get(1).execute(variantContainer) :
                        params.get(2).execute(variantContainer);
                break;
```

File: src/main/java/org/valdroz/vscript/BaseNode.java (hunk 10 above)
Comment (rpichaimani) [reply src/main/java/org/valdroz/vscript/BaseNode.java]: updated based on your comments. thanks

File: src/main/java/org/valdroz/vscript/Variant.java (hunk 12 above)
Comment (rpichaimani) [reply src/main/java/org/valdroz/vscript/Variant.java]: removed implementation.

File: src/main/java/org/valdroz/vscript/BaseNode.java (hunk 14 above)
Comment (rpichaimani) [reply src/main/java/org/valdroz/vscript/BaseNode.java:333]: updated based on your comments. thanks

File: src/main/java/org/valdroz/vscript/BaseNode.java (hunk 13 above)
Comment (rpichaimani) [reply src/main/java/org/valdroz/vscript/BaseNode.java]: updated based on your comments. thanks

**Review by valdroz — APPROVED** [review]

File: src/main/java/org/valdroz/vscript/BaseNode.java (hunk 11 above)
Comment (rpichaimani) [reply src/main/java/org/valdroz/vscript/BaseNode.java]: removed implementation.

**Pull request merged**